overlore/townhall/request.py

```
@@ -2,71 +2,120 @@
     realms = Realms.instance()
 
     ts = get_katana_timestamp(config.env["KATANA_URL"])
 
-    relevant_event: StoredEvent | None = events_db.fetch_most_relevant_event(realms.position_by_id(realm_id), ts)
-
-    if relevant_event is None:
-        return (0, None, None)
-
-    relevant_event_id: int = cast(int, relevant_event[0])
-
-    summary = vector_db.get_townhall_from_event(relevant_event_id, realm_id=realm_id)
-
-    return (relevant_event_id, relevant_event, summary)
+    return events_db.fetch_most_relevant_event(realms.position_by_id(realm_id), ts)
 
 
 def convert_dialogue_to_str(dialogue: list[DialogueSegment]) -> str:
     return "\n".join(
         [f"{dialogue_segment['full_name']}: {dialogue_segment['dialogue_segment']}" for dialogue_segment in dialogue]  # type: ignore[index]
     )
+
+
+def get_npc_thoughts(
+    npc_list: list[NpcEntity],
+    relevant_event: StoredEvent | None,
+    plotline: str | None,
+    llm: Llm,
+    townhall_db: TownhallDatabase,
+) -> list[str]:
+    thoughts = []
+    query = ""
+
+    if relevant_event is not None:
+        query += llm.nl_formatter.event_to_nl(relevant_event)
+    if plotline is not None:
+        query += plotline
+    if query != "":
+        query_embedding = llm.request_embedding(query)
+        for npc in npc_list:
+            # townhall_db.query_nearest_neighbour
+            res = townhall_db.query_cosine_similarity(query_embedding, npc["entity_id"])
+            if res is not None:
+                row_id = res[0]
+                similarity = res[1]
+
+                # Arbitrary number, can be tuned accordingly
+                if similarity >= 0.3:
+                    thoughts.append(npc["full_name"] + ":" + townhall_db.fetch_npc_thought(row_id))
+
+    return thoughts
+
+
+def get_entity_id_from_name(full_name: str, realm_npcs: list[NpcEntity]) -> int:
+    for npc in realm_npcs:
+        if full_name == npc["full_name"]:
+            return cast(int, npc["entity_id"])
+    return 0
```

Throw exception instead of returning 0

---

Done